From 2e55ecd2fa43f4b8fb134e54cad94169ff802037 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Thu, 16 Jan 2025 15:52:20 +0800
Subject: [PATCH 171/210] mtk: hostapd: Fix reassoc cb can't get STA error

Change mld reassoc resp frame sa to use mld address when the connection
is mlo.

Once MLD AP receive a non AP mld reassoc frame, kernel may not tanslate
the sa and da to mld address due to there is no recorded corresponding
STA entry. That makes mgmt->sa indicates a link address, which makes
handle_assoc_cb() can't find a related STA.

Remove unuse variable, mld_addrs_not_translated, which is added in commit,
e99670420 and the original logic is removed by commit, fd1a35e14.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 src/ap/ieee802_11.c | 15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 082b935de..c0f094e73 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5027,7 +5027,7 @@ static int add_associated_sta(struct hostapd_data *hapd,
 static u16 send_assoc_resp(struct hostapd_data *hapd, struct sta_info *sta,
 			   const u8 *addr, u16 status_code, int reassoc,
 			   const u8 *ies, size_t ies_len, int rssi,
-			   int omit_rsnxe, bool allow_mld_addr_trans)
+			   int omit_rsnxe)
 {
 	int send_len;
 	u8 *buf;
@@ -5436,8 +5436,7 @@ void fils_hlp_finish_assoc(struct hostapd_data *hapd, struct sta_info *sta)
 	reply_res = send_assoc_resp(hapd, sta, sta->addr, WLAN_STATUS_SUCCESS,
 				    sta->fils_pending_assoc_is_reassoc,
 				    sta->fils_pending_assoc_req,
-				    sta->fils_pending_assoc_req_len, 0, 0,
-				    true);
+				    sta->fils_pending_assoc_req_len, 0, 0);
 	os_free(sta->fils_pending_assoc_req);
 	sta->fils_pending_assoc_req = NULL;
 	sta->fils_pending_assoc_req_len = 0;
@@ -5539,7 +5538,6 @@ static void handle_assoc(struct hostapd_data *hapd,
 #endif /* CONFIG_FILS */
 	int omit_rsnxe = 0;
 	bool set_beacon = false;
-	bool mld_addrs_not_translated = false;
 	bool sae_pk = false;
 
 	if (len < IEEE80211_HDRLEN + (reassoc ? sizeof(mgmt->u.reassoc_req) :
@@ -5615,7 +5613,6 @@ static void handle_assoc(struct hostapd_data *hapd,
 			wpa_printf(MSG_DEBUG,
 				   "MLD: Switching to assoc hapd/station");
 			hapd = assoc_hapd;
-			mld_addrs_not_translated = true;
 		}
 	}
 #endif /* CONFIG_IEEE80211BE */
@@ -5998,11 +5995,9 @@ static void handle_assoc(struct hostapd_data *hapd,
 
 	if (resp >= 0)
 		reply_res = send_assoc_resp(hapd,
-					    mld_addrs_not_translated ?
-					    NULL : sta,
-					    mgmt->sa, resp, reassoc,
-					    pos, left, rssi, omit_rsnxe,
-					    !mld_addrs_not_translated);
+					    sta, sta ? sta->addr : mgmt->sa,
+					    resp, reassoc, pos, left, rssi,
+					    omit_rsnxe);
 	os_free(tmp);
 
 	/*
-- 
2.45.2

